semgrep-core/src/matching/Generic_vs_generic.ml

```diff
@@ -670,18 +673,36 @@ and m_expr_deep a b tin =
  *   - x = <expr>,
  *   - <call>(<exprs).
  * see SubAST_generic.subexprs_of_expr_implicit
+ *
+ * Also matches sub-expressions without decomposing, meaning it should match as the root.
  *)
 and m_expr_deep_implict a b tin =
   let symbolic_propagation = tin.config.Config.symbolic_propagation in
   let subexprs_of_expr =
     SubAST_generic.subexprs_of_expr_implicit ~symbolic_propagation
   in
-  m_deep m_expr_deep_implict m_expr subexprs_of_expr a b tin
-
+  m_deep m_expr_deep_implict m_expr_root subexprs_of_expr a b tin
+
+(* This is allows the matching of expressions, where we are considered to not be
+   at the "root" of the pattern.
+   This is because, for most of the other mutually recursive uses of `m_expr`, we are
+   usually having already decomposed on the pattern, meaning that it is safe to
+   use `m_expr`.
+*)
+and m_expr a b = m_expr_inner ~is_root:false a b
+
+(* This just calls `m_expr_inner` as the root. This is exposed, so that exterior uses will
+   properly register as the root, and should only be called at the start of matching.
+*)
+and m_expr_root a b = m_expr_inner ~is_root:true a b
+
+(* A wrapper for `m_expr`, which allows an `is_root` flag.
+   Any subsequent calls to `m_expr` will set it to false.
+*)
 (* coupling: if you add special sgrep hooks here, you should probably
  * also add them in m_pattern
  *)
-and m_expr a b =
+and m_expr_inner ~is_root a b =
```

Comment:
Nothing wrong with your solution really, just wanted to point out that you could have used an optional argument as in `m_expr ?(is_root=false)`. Did you avoid optional arguments for some reason in particular? Pure curiosity.

Reply:
the reason is that im silly and i started out with an optional argument, wrapped it with two functions which made the optional argument pointless, then eliminated the optional argument

Reply:
instead of just keeping the optional argument and keeping the name the same